tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank4TestHelpers.hpp

```diff
@@ -0,0 +1,241 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <iterator>
+#include <numeric>
+
+#include "DataStructures/Tensor/Expressions/Evaluate.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace TestHelpers::TensorExpressions {
+
+/// \ingroup TestingFrameworkGroup
+/// \brief Test that evaluating a right hand side tensor expression containing a
+/// single rank 4 tensor correctly assigns the data to the evaluated left hand
+/// side tensor
+///
+/// \details `TensorIndexA`, `TensorIndexB`, `TensorIndexC`, and  `TensorIndexD`
+/// can be any type of TensorIndex and are not necessarily `ti_a_t`, `ti_b_t`,
+/// `ti_c_t`, and `ti_d_t`. The "A", "B", "C", and "D" suffixes just denote the
+/// ordering of the generic indices of the RHS tensor expression. In the RHS
+/// tensor expression, it means `TensorIndexA` is the first index used,
+/// `TensorIndexB` is the second index used, `TensorIndexC` is the third index
+/// used, and `TensorIndexD` is the fourth index used.
+///
+/// If we consider the RHS tensor's generic indices to be (a, b, c, d), then
+/// this test checks that the data in the evaluated LHS tensor is correct
+/// according to the index orders of the LHS and RHS. The possible cases that
+/// are checked are when the LHS tensor is evaluated with index orders of all 24
+/// permutations of (a, b, c, d), e.g. (a, b, d, c), (a, c, b, d), ...
+///
+/// \tparam DataType the type of data being stored in the Tensors
+/// \tparam RhsSymmetry the ::Symmetry of the RHS Tensor
+/// \tparam RhsTensorIndexTypeList the RHS Tensor's typelist of
+/// \ref SpacetimeIndex "TensorIndexType"s
+/// \param tensorindex_a the first TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_a`
+/// \param tensorindex_b the second TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_B`
+/// \param tensorindex_c the third TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_c`
+/// \param tensorindex_d the fourth TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_D`
+template <typename DataType, typename RhsSymmetry,
+          typename RhsTensorIndexTypeList, typename TensorIndexA,
+          typename TensorIndexB, typename TensorIndexC, typename TensorIndexD>
+void test_evaluate_rank_4(const TensorIndexA& tensorindex_a,
```

Comment:
[not a change request, I just want to better understand] Why not use GENERATE_INSTANTIATIONS here to iterate over dims and frames as you've done in the lower-rank cases? Is it to avoid an explosion in the number of test cases and therefore test time?

Reply:
I think it was mostly just because when I was talking with @nilsdeppe weeks ago about test cases, we felt that if we exhaustively tested rank 3 tensors, there wasn't much of a need to exhaustively test rank 4 tensors since most of the edge cases you can think of with reordering symmetries and indices are already addressed by exhaustively testing rank 3. But, of course, exhaustively testing rank 4 is safer and more thorough.

Let me know if you would like me to implement the same exhaustive testing for rank 4 that I did with the others!

Reply:
No, I agree - I think you're in the diminishing returns territory here, where the test has to get a lot more complicated to test any potential remain edge cases. I think you can leave the code unchanged.

However, it might be useful to add a short note to the documentation to make the intent clear? Perhaps something like (I'm mostly reusing your words here) "The rank3 case is exhaustively tested, and should test most edge cases with reordered indices and symmetries. So we test the rank4 case less exhaustively to keep the test manageable"

Reply:
Diminishing returns is a good way to describe it - that's exactly it. We didn't think there was too much to gain. I'll add a note! I am a supporter of making things clear.